docs(store): add docs for createActionGroup api (#3530)
ngrx/platform: projects/ngrx.io/content/guide/store/action-groups.md

# Action Groups

Action groups is a feature that allows you to group actions together. While writing actions, the actions in most of the cases looks like below.

```typescript
import { createAction, props } from '@ngrx/store';

export const opened = createAction('[Products Page] Opened');

// defining an action with payload using the `props` function
export const paginationChanged = createAction(
  '[Products Page] Pagination Changed',
  props<{ page: number; offset: number }>()
);

// defining an action with payload using the props factory
export const queryChanged = createAction(
  '[Product Page] Query Changed',
  (query: Error) => ({ query })
);

> ```

Author: Done.

With the help of this new `createActionGroup` api, we can avoid the above code and write the following code.

> ```typescript
> 
> import { createActionGroup } from '@ngrx/store';
> 
> export const productPageActions = createActionGroup({
>   source: '[Product Page]',
>   events: {
>     'Pagination Changed': props<{ page: number; offset: number }>(),
>     'Query Changed': (query: Error) => ({ query }),
>   }
> });
> 
> ```

Reviewer: Square brackets are not needed for the action group source. I also suggest adding the missing `Opened` action and comments about events:

```suggestion
<code-example header="products-page.actions.ts">
import { createActionGroup, emptyProps, props } from '@ngrx/store';

export const ProductsPageActions = createActionGroup({
  source: 'Products Page',
  events: {
    // defining an event without payload using the `emptyProps` function
    'Opened': emptyProps(),
    
    // defining an event with payload using the `props` function
    'Pagination Changed': props&lt;{ page: number; offset: number }&gt;(),
    
    // defining an event with payload using the props factory
    'Query Changed': (query: string) => ({ query }),
  }
});
</code-example>
```

Author: Done

After moving to the new `createActionGroup` api, we can use the following code to import and dispatch the action.

```typescript

> import { productPageActions } from './product.actions';
> 
> productPageActions.paginationChanged({ page: 1, offset: 10 });
> productPageActions.queryChanged(new Error('Query is invalid'));

Reviewer: Here we can also add an action type for each generated action:

```suggestion
import { ProductsPageActions } from './products-page.actions';

// action type: [Products Page] Opened
ProductsPageActions.opened();

// action type: [Products Page] Pagination Changed
ProductsPageActions.paginationChanged({ page: 10, offset: 100 });

// action type: [Products Page] Query Changed
ProductsPageActions.queryChanged('ngrx');
```

Author: Thanks 😊 done with the changes.

# Action Groups

> Action groups is a feature that allows you to group actions together. While writing actions, the actions in most of the cases looks like below.

Reviewer 2: ```suggestion
Action groups is a feature to group actions together. 
While writing actions, the actions in most of the cases looks like below.
```

<code-example header="products-page.actions.ts">
import { createAction, props } from '@ngrx/store';

> // defining an action without payload
> export const opened = createAction('[Products Page] Opened');
> // defining an action with payload using the `props` function
> export const paginationChanged = createAction(
>   '[Products Page] Pagination Changed',
>   props&lt;{ page: number; offset: number }&gt;()
> );
> // defining an action with payload using the props factory
> export const queryChanged = createAction(
>   '[Products Page] Query Changed',
>   (query: string) => ({ query })
> );

Reviewer 2: I think we can leave out the comments to explain how an action is created, thoughts?

```suggestion
export const opened = createAction('[Products Page] Opened');

export const paginationChanged = createAction(
  '[Products Page] Pagination Changed',
  props&lt;{ page: number; offset: number }&gt;()
);

export const queryChanged = createAction(
  '[Products Page] Query Changed',
  (query: string) => ({ query })
);
```

Author: makes sense.

</code-example>

> With the help of this new `createActionGroup` api, we can avoid the above code and write the following code.

Reviewer 2: The docs shouldn't mention if it's new or old.

```suggestion
With the help of the `createActionGroup` API, we can avoid the above code and write the following code.
```

Reviewer 2: Should we mention here why/when `createActionGroup` should be used?
E.g. to group actions that have the same source, making the code more compact, etc, etc

<code-example header="products-page.actions.ts">

import { createActionGroup, emptyProps, props } from '@ngrx/store';
export const ProductsPageActions = createActionGroup({
  source: 'Products Page',
  events: {
    // defining an event without payload using the `emptyProps` function
    'Opened': emptyProps(),
    
    // defining an event with payload using the `props` function
    'Pagination Changed': props&lt;{ page: number; offset: number }&gt;(),
    
    // defining an event with payload using the props factory
    'Query Changed': (query: string) => ({ query }),
  }
});
</code-example>

> After moving to the new `createActionGroup` api, we can use the following code to import and dispatch the action.

Reviewer 2: ```suggestion
To dispatch an action from the group, import the group and invoke an action.
This returns an action that you can then dispatch to the store.
```

This snippet should also show how to dispatch an action to the store.
The current snippet only invokes the action, but doesn't dispatch it.

```typescript

import { ProductsPageActions } from './products-page.actions';
// action type: [Products Page] Opened
ProductsPageActions.opened();
// action type: [Products Page] Pagination Changed
ProductsPageActions.paginationChanged({ page: 10, offset: 100 });
// action type: [Products Page] Query Changed
ProductsPageActions.queryChanged('ngrx');

```

> You dont need to use below syntax to import actions.
> 
> ```typescript 
> 
> import * as productActions from './product.actions';
> 
> ```

Reviewer 2: What do you mean here?
I think it can be removed.

```suggestion
You dont need to use below syntax to import actions.

```typescript 

import * as productActions from './product.actions';

```
```

# Action Groups

Action groups is a feature to group actions together. While writing actions, the actions in most of the cases looks like below.

<code-example header="products-page.actions.ts">
import { createAction, props } from '@ngrx/store';

export const opened = createAction('[Products Page] Opened');

export const paginationChanged = createAction(
  '[Products Page] Pagination Changed',
  props&lt;{ page: number; offset: number }&gt;()
);

export const queryChanged = createAction(
  '[Products Page] Query Changed',
  (query: string) => ({ query })
);
</code-example>

With the help of the `createActionGroup` api, we can avoid the above code and write the following code. This will create an action group with the name `ProductsPageActions` and the actions will be grouped together. It is very useful when we want to group actions related to same source `[Products Page]` in the example together. It makes the code more compact.

> <code-example header="products-page.actions.ts">
> 
> import { createActionGroup, emptyProps, props } from '@ngrx/store';
> export const ProductsPageActions = createActionGroup({

Reviewer: ```suggestion
<code-example header="products-page.actions.ts">
import { createActionGroup, emptyProps, props } from '@ngrx/store';

export const ProductsPageActions = createActionGroup({
```

  source: 'Products Page',
  events: {
    // defining an event without payload using the `emptyProps` function
    'Opened': emptyProps(),
    
    // defining an event with payload using the `props` function
    'Pagination Changed': props&lt;{ page: number; offset: number }&gt;(),
    
    // defining an event with payload using the props factory
    'Query Changed': (query: string) => ({ query }),
  }
});
</code-example>

To dispatch an action from the group, import the group and invoke an action.
This returns an action that you can then dispatch to the store.

```typescript

import { ProductsPageActions } from './products-page.actions';

@Component({ /* ... */ })
export class ProductsComponent implements OnInit {
  constructor(private readonly store: Store) {}

>   ngOnInit(): void {
>     // action type: [Products Page] Opened
>     ProductsPageActions.opened();
>     // action type: [Products Page] Pagination Changed
>     ProductsPageActions.paginationChanged({ page: 10, offset: 100 });
>     // action type: [Products Page] Query Changed
>     ProductsPageActions.queryChanged('ngrx');
>   }

Reviewer: ```suggestion
  ngOnInit(): void {
    // action type: [Products Page] Opened
    this.store.dispatch(ProductsPageActions.opened());
  }
  
  onPaginationChanged(page: number, offset: number): void {
    // action type: [Products Page] Pagination Changed
    this.store.dispatch(ProductsPageActions.paginationChanged({ page, offset }));
  }
  
  onQueryChanged(query: string): void {
    // action type: [Products Page] Query Changed
    this.store.dispatch(ProductsPageActions.queryChanged(query));
  }
```

}

```

## Limitations and Restrictions

The action names are generated from the event names and camel-cased, with `createAction` you can can use action creator name as the event name. For example in below code snippet the creator name will be `productsLoadedSuccess` but with `createActionGroup` the creator name would be `productsAreLoadedSuccessfully`.

> ```typescript
> 
> export const productsLoadedSuccess = createAction(
>   '[Products API] Products Are Loaded Successfully',
>   props<{ products: Product[] }>()
> );
> 
> ```
> 
> The `createActionGroup` function allows using letters, numbers, and spaces to define the event name. But below symbols are not allowed.
> 
> ```typescript
> 
> type ForbiddenCharactersStr =
>   '/ \\ | < > [ ] { } ( ) . , ! ? # % ^ & * + - ~ \' " `';
> 
> ```

Reviewer 2: ```suggestion
```
